ecosystem/indexer-grpc/indexer-grpc-table-info/src/runtime.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    counters::{
+        IndexerTableInfoStep, DURATION_IN_SECS, LATEST_PROCESSED_VERSION, NUM_TRANSACTIONS_COUNT,
+        SERVICE_TYPE,
+    },
+    table_info_parser::TableInfoParser,
+};
+use aptos_api::context::Context;
+use aptos_config::config::NodeConfig;
+use aptos_logger::{error, info};
+use aptos_mempool::MempoolClientSender;
+use aptos_moving_average::MovingAverage;
+use aptos_storage_interface::DbReaderWriter;
+use aptos_types::chain_id::ChainId;
+use std::{env, sync::Arc};
+use tokio::runtime::Runtime;
+
+/// Creates a runtime which creates a thread pool which sets up fullnode indexer table info service
+/// Returns corresponding Tokio runtime
+pub fn bootstrap(
+    config: &NodeConfig,
+    chain_id: ChainId,
+    db: DbReaderWriter,
+    mp_sender: MempoolClientSender,
+) -> Option<Runtime> {
+    if !config.indexer_table_info.enabled {
+        return None;
+    }
+
+    let runtime = aptos_runtimes::spawn_named_runtime("table-info".to_string(), None);
+
+    let node_config = config.clone();
+    let parser_task_count = node_config.indexer_table_info.parser_task_count;
+    let parser_batch_size = node_config.indexer_table_info.parser_batch_size;
+
+    // Compute to get the start version to start parsing.
+    // Take the maximum of (0, next version from rocksdb - # of transactions processed per loop)
+    let next_version = db.reader.get_indexer_async_v2_next_version().unwrap();
+    let subtraction_value = parser_batch_size as u64 * parser_task_count as u64;
+    let start_version = if next_version > subtraction_value {
+        next_version - subtraction_value
+    } else {
+        0
+    };
+
+    // Spawn the runtime for table info parsing
+    runtime.spawn(async move {
+        let context =
+            Arc::new(Context::new(chain_id, db.reader.clone(), mp_sender, node_config));
+        let mut ma = MovingAverage::new(10_000);
+        let mut base: u64 = 0;
+        let db = db.clone();
+        let mut parser =
+            TableInfoParser::new(context, start_version, parser_task_count, parser_batch_size);
```

Comment:
I think that you should either move all this logic into TableInfoParser (it becomes essentially the coordinator), or kill TableInfoParser and move the logic here. As it stand you have some business logic here and some in TableInfoParser which makes the code confusing. 

The overall logic should also be simplified. Currently you track some type of end version and max version in multiple places. All you need is this: 
let mut batch_start = bla;
loop {
  let batch_end = min(ledger_version, batch_size*batch_count);
  let batches = chunk(batch_start, batch_end) // vector of (start_version, end_version)
  for batch in batches: 
     // create workers
  tasks.join
  batch_start = final_end_version + 1; // or fetch from rocksdb if it's available. 
}